Sources/SyncDataProviders/Settings/SettingsProvider.swift

```diff
@@ -0,0 +1,341 @@
+                        context: context,
+                        crypter: crypter,
+                        deduplicateEntities: isInitial
+                    )
+                    let idsOfItemsToClearModifiedAt = try cleanUpSentItems(
+                        sent,
+                        receivedKeys: Set(responseHandler.receivedByKey.keys),
+                        clientTimestamp: clientTimestamp,
+                        in: context
+                    )
+                    try responseHandler.processReceivedSettings()
+
+#if DEBUG
+                    try willSaveContextAfterApplyingSyncResponse()
+#endif
+                    let keys = idsOfItemsToClearModifiedAt.union(
+                        Set(responseHandler.receivedByKey.keys)
+                            .subtracting(responseHandler.idsOfItemsThatRetainModifiedAt)
+                    )
+                    try clearModifiedAtAndSaveContext(keys: keys, clientTimestamp: clientTimestamp, in: context)
+                    break
+                } catch {
+                    if (error as NSError).code == NSManagedObjectMergeError {
+                        context.reset()
+                        saveAttemptsLeft -= 1
+                        if saveAttemptsLeft == 0 {
+                            saveError = error
+                            break
+                        }
+                    } else {
+                        saveError = error
+                        break
+                    }
+                }
+
+            }
+        }
+        if let saveError {
+            throw saveError
+        }
+
+        if let serverTimestamp {
+            lastSyncTimestamp = serverTimestamp
+            syncDidUpdateData()
+        }
+    }
+
+    func cleanUpSentItems(
+        _ sent: [Syncable],
+        receivedKeys: Set<String>,
+        clientTimestamp: Date,
+        in context: NSManagedObjectContext
+    ) throws -> Set<String> {
+
+        if sent.isEmpty {
+            return []
+        }
+        let keys = sent.compactMap { SyncableSettingAdapter(syncable: $0).uuid }
+        let settingsMetadata = try SyncableSettingsMetadataUtils.fetchSettingsMetadata(for: keys, in: context)
+
+        var idsOfItemsToClearModifiedAt = Set<String>()
+
+        for metadata in settingsMetadata {
+            guard let handler = settingsHandlers[Setting(key: metadata.key)] else {
+                continue
+            }
+
+            if let lastModified = metadata.lastModified, lastModified > clientTimestamp {
+                continue
+            }
+            let isLocalChangeRejectedBySync: Bool = receivedKeys.contains(metadata.key)
+            let isPendingDeletion = try handler.getValue() == nil
+            if isPendingDeletion, !isLocalChangeRejectedBySync {
+                try handler.setValue(nil)
+                context.delete(metadata)
+            } else {
+                context.delete(metadata)
```

Comment:
Is this delete intentional? Wouldn't that affect any update, not only deletes?

Reply:
This function clears timestamps from all sent items. In settings, clearing timestamp is equivalent to deleting its respective metadata object.

It looks like this for bookmarks:
```swift
if bookmark.isPendingDeletion, !isLocalChangeRejectedBySync {
    context.delete(bookmark)
} else {
    bookmark.modifiedAt = nil
    idsOfItemsToClearModifiedAt.insert(uuid)
}
```

I should put `context.delete(metadata)` outside of the `if` statement because it's a common part of deleting an object and clearing a timestamp on a locally updated object.

Reply:
I see. So metadata is only present if there's something to send right? It probably would be simpler to just always keep one instance around (and update timetamp if needed) but that approach is also fine. Since we are removing/adding objects I think:
- it is posible that setting new timestamp value will fail (metadata object could be deleted by sync)
- It is possible we have more that one metadata created (as there is no synchornization on which context or logic flow could lead to creation of metadata).

Both are unlikely, but would be good to know if that happens.

Reply:
Yes, the idea was to delete metadata objects when there's nothing to sync.

Setting new timestamp value should always work because if the object wasn't present, we create one (see [this function](https://github.com/duckduckgo/BrowserServicesKit/blob/dominik/sync-settings/Sources/DDGSync/SyncableSettingsMetadata.swift#L65-L85)). This can of course lead to creation of multiple objects per key. Let me see if I can rewrite it to keep a single metadata object around.

Reply:
I've updated logic to keep metadata objects around (once they're created in `prepareForFistSync`), and only clear timestamps after sync. I added `SyncError.settingsMetadataNotPresent` and a self-healing mechanism of creating a new object in this scenario.

Please have a look at https://github.com/duckduckgo/BrowserServicesKit/pull/485/commits/8de5b4aa22c6b54bd410c2de56666ccf20ae2f07.